File: cs/sample/ab/ab.cs
using Etsm;

// Output: " ->A  A->  ->B "
Foo foo = new Foo();
foo.Run();

public class Foo
{
    private StateMachine<State> stateMachine;
    private State stateA;
    private State stateB;

    public Foo()
    {
        stateMachine = new StateMachine<State>();
        stateA = new State(EnterA, ExitA);
        stateB = new State(EnterB, null);
    }

    public void Run()
    {
        stateMachine.Transition(stateA);
        stateMachine.Transition(stateB);
        stateMachine.Transition(null);
    }

    void EnterA()
    {
        System.Console.Write(" ->A ");
    }

    void ExitA()
    {
        System.Console.Write(" A-> ");
    }

    void EnterB()
    {
        System.Console.Write(" ->B ");
    }
}

File: c++/sample/ab/ab.cpp
#include <etsm.h>
#include <iostream>

using namespace etsm;

class Foo
{
public:
	Foo()
		: sm(this),
		a(&Foo::EnterA, &Foo::ExitA),
		b(&Foo::EnterB, nullptr)
	{
	}

	void Run()
	{
		sm.Transition(&a);
		sm.Transition(&b);
		sm.Transition(nullptr);
		std::cout << output << std::endl;
	}

private:
	StateMachine<Foo> sm;
	State<Foo> a;
	State<Foo> b;
	std::string output;

	void EnterA() { output += " ->A "; }
	void ExitA() { output += " A-> "; }
	void EnterB() { output += " ->B "; }
};

// Output: " ->A  A->  ->B "
int main()
{
	Foo foo;
	foo.Run();
}
